fvtest/compilertriltest/MinimalTest.cpp

```diff
@@ -0,0 +1,372 @@
+
+   DefineName("SubArgArg");
+   DefineParameter("arg1", Int32);
+   DefineParameter("arg2", Int32);
+   DefineReturnType(Int32);
+   }
+
+bool
+SubArgArgMethod::buildIL()
+   {
+   Return(
+      Sub(
+         Load("arg1"),
+         Load("arg2")));
+   return true;
+   }
+
+class FactorialMethod : public TR::MethodBuilder
+   {
+   public:
+   FactorialMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+FactorialMethod::FactorialMethod(TR::TypeDictionary *types)
+   : TR::MethodBuilder(types, NULL)
+   {
+   DefineLine(LINETOSTR(__LINE__));
+   DefineFile(__FILE__);
+
+   DefineName("Factorial");
+   DefineParameter("x", Int32);
+   DefineReturnType(Int32);
+   }
+
+bool
+FactorialMethod::buildIL()
+   {
+   TR::IlBuilder *baseCase=NULL, *recursiveCase=NULL;
+   IfThenElse(&baseCase, &recursiveCase,
+      LessThan(
+         Load("x"),
+         ConstInt32(1)));
+
+   DefineLocal("result", Int32);
+
+   baseCase->Store("result",
+   baseCase->   ConstInt32(1));
+   recursiveCase->Store("result",
+   recursiveCase->   Mul(
+   recursiveCase->      Call("Factorial", 1,
+   recursiveCase->         Sub(
+   recursiveCase->            Load("x"),
+   recursiveCase->            ConstInt32(1))),
+   recursiveCase->      Load("x")));
+
+   Return(
+      Load("result"));
+   return true;
+   }
+
+class RecursiveFibonnaciMethod : public TR::MethodBuilder
+   {
+   public:
+	RecursiveFibonnaciMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+
+RecursiveFibonnaciMethod::RecursiveFibonnaciMethod(TR::TypeDictionary *types)
+   : TR::MethodBuilder(types, NULL)
+   {
+   DefineLine(LINETOSTR(__LINE__));
+   DefineFile(__FILE__);
+
+   DefineName("fib_recur");
+   DefineParameter("n", Int32);
+   DefineReturnType(Int32);
+
+   }
+
+bool
+RecursiveFibonnaciMethod::buildIL()
+   {
+   TR::IlBuilder *baseCase=NULL, *recursiveCase=NULL;
+   IfThenElse(&baseCase, &recursiveCase,
+      LessThan(
+         Load("n"),
+         ConstInt32(2)));
+
+   DefineLocal("result", Int32);
+
+   baseCase->Store("result",
+   baseCase->   Load("n"));
+   recursiveCase->Store("result",
+   recursiveCase->   Add(
+   recursiveCase->      Call("fib_recur", 1,
+   recursiveCase->         Sub(
+   recursiveCase->            Load("n"),
+   recursiveCase->            ConstInt32(1))),
+   recursiveCase->      Call("fib_recur", 1,
+   recursiveCase->         Sub(
+   recursiveCase->            Load("n"),
+   recursiveCase->            ConstInt32(2)))));
+
+   Return(
+      Load("result"));
+   return true;
+   }
+
+
+template <typename F>
+F compile(TR::MethodBuilder *builder) {
+	TR::ResolvedMethod resolvedMethod(builder);
+	TR::IlGeneratorMethodDetails details(&resolvedMethod);
+	int32_t rc = 0;
+	F entry = (F)compileMethodFromDetails(NULL, details, warm, rc);
+	return entry;
+}
+
+#define COMPILE(test, name) { \
+   TR::TypeDictionary types; \
+   name ## Method builder(&types); \
+   entry = compile<decltype(entry)>(&builder); \
+   }
```

Comment:
I'm personally biased against using macros, and I would prefer if you just wrote this code out by hand, or used helper functions to reduce the boilerplate.

Reply:
@rwy0717: I rewrote it using templates and it looks much nicer to me. Thanks for the suggestion!